tacs/constraints/panel_width.py

```diff
@@ -0,0 +1,422 @@
+
+        lower: float or complex
+            lower bound for constraint. Not used.
+
+        upper: float or complex
+            upper bound for constraint. Not used.
+
+        dvIndex : int
+            Index number of the panel length DV's. Defaults to 0.
+
+        """
+        if compIDs is not None:
+            # Make sure CompIDs is flat and get element numbers on each proc corresponding to specified compIDs
+            compIDs = self._flatten(compIDs)
+        else:
+            nComps = self.meshLoader.getNumComponents()
+            compIDs = range(nComps)
+
+        dvGlobalInds = []
+        dvProcs = []
+        dvLocalInds = []
+        boundaryNodeGlobalInds = []
+        boundaryNodeLocalInds = []
+        boundaryNodeLocalProcs = []
+        refAxes = []
+        widthAxes = []
+        dvJacRows = []
+        dvJacCols = []
+        dvJacVals = []
+        coordJacRows = []
+        coordJacCols = []
+        for ii in range(self.comm.size):
+            dvJacRows.append([])
+            dvJacCols.append([])
+            dvJacVals.append([])
+            coordJacRows.append([])
+            coordJacCols.append([])
+
+        # Get the boundary node IDs for each component
+        boundaryNodeIDs, boundaryNodeCoords = self._getComponentBoundaryNodes(compIDs)
+
+        if self.rank == 0:
+            constraintInd = 0
+            for compID in compIDs:
+                # Get the TACS element object associated with this compID to
+                # get the ref axis
+                elemObj = self.meshLoader.getElementObject(compID, 0)
+                transObj = elemObj.getTransform()
+                try:
+                    refAxis = transObj.getRefAxis()
+                except AttributeError as e:
+                    raise AttributeError(
+                        f"The elements in component {self.meshLoader.compDescripts[compID]} do not have a reference axis. Please define one by using the 'ShellRefAxisTransform' class with your elements"
+                    ) from e
+
+                # For a more accurate length calculation, roject the ref axis
+                # onto the "average" plane of the baseline panel geometry by
+                # using an SVD to compute a normal vector
+                centroid = np.mean(boundaryNodeCoords[compID], axis=0, keepdims=True)
+                centredPoints = boundaryNodeCoords[compID] - centroid
+                _, _, VT = np.linalg.svd(centredPoints, full_matrices=False)
+                panelNormal = VT[-1]
+                refAxis -= np.dot(refAxis, panelNormal) * panelNormal
+                refAxis /= np.linalg.norm(refAxis)
+                refAxes.append(refAxis)
+
+                # now this refAxis is the panel length axis, compute the width axis which is normal to the length + panel normal axis
+                widthAxis = np.cross(refAxis, panelNormal)
+                widthAxis /= np.linalg.norm(widthAxis)
+                widthAxes.append(widthAxis)
```

Comment:
These are the only lines that actually need to be different in the panel width constraint compared to the panel length constraint. I think we can reduce the amount of code in this class massively by putting this in a function called something like `computeRefAxis` just having the panel width constraint override that. Then there'd be no need to override `evalConstraints` and `evalConstraintSens` either.

I'm imagining something like this:

```python
refAxes.append(self.computeRefAxis(boundaryNodeCoords[compID], refAxis)
```

Reply:
Yeah good point, just cleaned it up